### PR TITLE
[Feature](tvf) Support cdc stream tvf for mysql and pg

### PR DESCRIPTION
**Pull request by JNSimba**
### What problem does this PR solve?

#58898 #59461 These two features support consuming data from MySQL and PostgreSQL. 

Based on this, this PR introduces cdc_stream tvf, which allows consuming data from MySQL and PostgreSQL in TVF format.

Example
```
select * from cdc_stream(
    "type" = "mysql",
    "jdbc_url" = "jdbc:mysql://127.0.0.1:3306",
    "driver_url" = "mysql-connector-j-8.0.31.jar",
    "driver_class" = "com.mysql.cj.jdbc.Driver",
    "user" = "root",
    "password" = "123456",
    "database" = "test_cdc_db",
    "table" = "user_info",
    "offset" = '{"file":"binlog.000003","pos":"496"}'
)
```

### Release note

None

### Check List (For Author)

- Test 
    - [x] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by hello-stephen** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

**Comment by hello-stephen** [comment]
# BE UT Coverage Report
Increment line coverage `0.00% (0/79)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/d84fcddbe4aa9142e912da69ab6bafe20075ca2e_d84fcddbe4aa9142e912da69ab6bafe20075ca2e/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/d84fcddbe4aa9142e912da69ab6bafe20075ca2e_d84fcddbe4aa9142e912da69ab6bafe20075ca2e/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 52.41% (19093/36431) |
| Line Coverage     | 35.77% (177218/495488) |
| Region Coverage   | 32.25% (137009/424848) |
| Branch Coverage   | 33.17% (59277/178717) |

File: fs_brokers/cdc_client/src/main/java/org/apache/doris/cdcclient/source/reader/mysql/MySqlSourceReader.java [hunk 1]
```diff
@@ -170,39 +169,14 @@ public SplitReadResult readSplitRecords(JobBaseRecordRequest baseReq) throws Exc
...
-                    currentSplitRecords.getSplitId());
-        }
+        // Create a new snapshot/binlog split based on offset and start the reader.
+        LOG.info("create new split reader for {} with offset {}", baseReq.getJobId(), offsetMeta);
+        // build split
+        Tuple2<MySqlSplit, Boolean> splitFlag = createMySqlSplit(offsetMeta, baseReq);
+        MySqlSplit split = splitFlag.f0;
+        // it's necessary to ensure that the binlog reader is already closed.
+        this.currentSplitRecords = pollSplitRecordsWithSplit(split, baseReq);
+        this.currentSplit = split;
```
[the remark this answers is not in the document]
Comment (JNSimba) [reply fs_brokers/cdc_client/src/main/java/org/apache/doris/cdcclient/source/reader/mysql/MySqlSourceReader.java]: Currently, the reader is closed in finishSplitRecords after each read operation, so the other branches here are probably useless.

**Comment by JNSimba** [comment]
run buildall

**Comment by JNSimba** [comment]
run buildall

**Comment by hello-stephen** [comment]
# FE UT Coverage Report
Increment line coverage `  3.08% (2/65)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/60116_e02811eaefda6a3a1631ee0683594c271015510b/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/60116_e02811eaefda6a3a1631ee0683594c271015510b/fe_report/index.html)

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `65.82% (52/79)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/60116_e02811eaefda6a3a1631ee0683594c271015510b_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/60116_e02811eaefda6a3a1631ee0683594c271015510b_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 73.12% (26118/35719) |
| Line Coverage     | 56.20% (278285/495144) |
| Region Coverage   | 54.08% (232333/429612) |
| Branch Coverage   | 55.65% (99946/179604) |

**Comment by hello-stephen** [comment]
# FE Regression Coverage Report
Increment line coverage `  3.08% (2/65)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/60116_e02811eaefda6a3a1631ee0683594c271015510b_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/60116_e02811eaefda6a3a1631ee0683594c271015510b_merge_fe/report/index.html)

**Review by liaoxin01 — DISMISSED** [review]
LGTM

**Review by liaoxin01 — DISMISSED** [review]
LGTM

**Review by dataroaring — COMMENTED** [review]
## Review: `cdc_stream` TVF for MySQL/PostgreSQL

Nice approach reusing `http_file_reader` as the streaming CDC transport — clever to treat the CDC stream as a remote JSON file over HTTP. A few concerns below, roughly by severity.

---

### Critical

**1. Brittle TVF-vs-Job discrimination via `isLong()`**

```java
// PipelineCoordinator.java
if (!isLong(fetchRecord.getJobId())) {
    sourceReader.close(fetchRecord);
    Env.getCurrentEnv().close(fetchRecord.getJobId());
}
```

Using "can the jobId be parsed as a Long?" as an implicit contract to decide cleanup behavior will break the moment anyone changes the ID format. Suggest an explicit field (e.g., `boolean isTvfRequest` or a `RequestSource` enum).

**2. `Long.parseLong(jobId)` in `DorisBatchStreamLoad.commitOffset`**

After migrating `jobId` from `Long` to `String` everywhere, this line will throw `NumberFormatException` if a UUID-based TVF jobId ever reaches this path:

```java
// DorisBatchStreamLoad.java
.jobId(Long.parseLong(jobId))
```

Even if the current flow prevents it, this is a latent crash.

**3. `getRemoteDbName` silently changed from checked to unchecked exception**

```java
// Before: throws JobException
// After:  throws RuntimeException
```

Existing callers catching `JobException` will now miss this exception entirely. This is a silent behavior change.

**4. Missing `database` validation in `CdcStreamTableValuedFunction.validate()`**

`jdbc_url`, `type`, `table`, `offset` are validated — but `database` is not, even though `getRemoteDbName()` requires it for MySQL. A user omitting `"database"` will get a confusing NPE instead of a clear error message.

**5. `taskOffsetCache` unbounded memory leak**

```java
private final Map<String, Map<String, String>> taskOffsetCache = new ConcurrentHashMap<>();
```

Entries are only removed when `getOffsetWithTaskId()` is called. If a client crashes or never fetches the offset, entries accumulate forever. Consider a TTL-based eviction (e.g., Guava `Cache` with `expireAfterWrite`) or a size bound.

---

### Major

**6. Streaming loop can run forever**

The idle timeout (`POLL_SPLIT_RECORDS_TIMEOUTS`) only applies when the source produces no records. A continuously-producing source keeps the loop (and the HTTP response) open indefinitely, meaning the SELECT query never completes. Consider adding `max_records` or `max_duration` parameters.

**7. `BufferedOutputStream` not closed in finally**

```java
BufferedOutputStream bos = new BufferedOutputStream(rawOutputStream);
try {
    // ... write records ...
    bos.flush();
} finally {
    sourceReader.commitSourceOffset(...);  // cleanup only — bos not closed
}
```

If an exception occurs between writes and the explicit `flush()`, buffered data could be lost.

**8. Password in HTTP payload**

`recordRequest.setConfig(properties)` includes the plaintext password. Even on `127.0.0.1`, this will appear in request logging, debug dumps, or error messages that include the payload. Consider filtering sensitive fields before serialization.

---

### Minor

- **`http_file_reader.cpp` CDC coupling**: The generic file reader now manages CDC process lifecycle and port discovery. Consider a pre-processing hook or specialized subclass to keep the I/O layer clean.
- **File size `Integer.MAX_VALUE`**: `new TBrokerFileStatus(URI, false, Integer.MAX_VALUE, false)` — if scan planning uses this for memory allocation or split decisions, it could cause issues.
- **`getServerId` hash collision**: `jobId.hashCode() & Integer.MAX_VALUE` is a good fix for the `Math.abs(MIN_VALUE)` edge case, but MySQL server-id collisions across concurrent CDC sessions could cause silent data issues. Consider a better hash or monotonic allocator.
- **Postgres test doesn't assert content**: The result is logged but never asserted — only verifies no exception.
- **Inconsistent `ObjectMapper`**: `PipelineCoordinator.objectMapper` is non-final and mutable; should be `private static final` like in `CdcStreamTableValuedFunction`.

**Review by dataroaring — APPROVED** [review]
LGTM

**Review by liaoxin01 — APPROVED** [review]
LGTM
